packages/shared/src/utils/error.ts

```diff
@@ -1,8 +1,100 @@
 import type { ZodIssue } from 'zod';
 
+/**
+ * Determines if a path should be simplified in error messages.
+ *
+ * Simplifies paths that are:
+ * - Array element validations (e.g., [0, 'rules'] becomes just the message)
+ * - Top-level custom validations with no meaningful path context
+ *
+ * Business rationale: Array indices like "0.rules" are confusing to end users
+ * because they don't know which item in the array failed, and the custom error
+ * message typically provides better context (e.g., "Expiry rule is required").
+ *
+ * @param path - The Zod error path segments.
+ * @param code - The Zod error code.
+ * @returns True if the path should be simplified (message only), false otherwise.
+ */
+const shouldSimplifyPath = (
+  path: (string | number)[],
+  code: string,
+): boolean => {
+  // Only simplify custom validation errors
+  if (code !== 'custom') {
+    return false;
+  }
+
+  // Empty path - already just the message
+  if (path.length === 0) {
+    return true;
+  }
+
+  // Pattern: [arrayIndex, fieldName] - e.g., [0, 'rules']
+  // This represents a refinement on an array element's field
+  if (path.length === 2 && typeof path[0] === 'number') {
+    return true;
+  }
+
+  // Pattern: [arrayIndex, fieldName, nestedIndex] - e.g., [0, 'rules', 1]
+  // This represents a refinement on nested array elements
+  if (
+    path.length === 3 &&
+    typeof path[0] === 'number' &&
+    typeof path[2] === 'number'
+  ) {
+    return true;
+  }
```

Comment:
nit: is it possible for this to nest even further? Should this be a programmatic check?